arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/Fold.kt

```diff
@@ -108,27 +108,92 @@ public inline fun <R, A, B> fold(
   }
 }
 
+/**
+ * Inspect a [Trace] value of [R].
+ *
+ * Tracing [R] can be useful to know where certain errors, or failures are coming from.
+ * Let's say you have a `DomainError`, but it might be raised from many places in the project.
+ *
+ * You would have to manually _trace_ where this error is coming from,
+ * instead [Trace] offers you ways to inspect the actual stacktrace of where the raised value occurred.
+ *
+ * Beware that tracing can only track the [Raise.bind] or [Raise.raise] call that resulted in the [R] value,
+ * and not any location of where the [R], or [Either.Left] value was created.
+ *
+ * ```kotlin
+ * public fun main() {
+ *   val error = effect<String, Int> { raise("error") }
+ *   error.traced { (trace, _: String) -> trace.printStackTrace() }
+ *     .fold({ require(it == "error") }, { error("impossible") })
+ * }
+ * ```
+ * ```text
+ * arrow.core.continuations.RaiseCancellationException: Raised Continuation
+ *   at arrow.core.continuations.DefaultRaise.raise(Fold.kt:77)
+ *   at MainKtKt$main$error$1.invoke(MainKt.kt:6)
+ *   at MainKtKt$main$error$1.invoke(MainKt.kt:6)
+ *   at arrow.core.continuations.Raise$DefaultImpls.bind(Raise.kt:22)
+ *   at arrow.core.continuations.DefaultRaise.bind(Fold.kt:74)
+ *   at arrow.core.continuations.Effect__TracingKt$traced$2.invoke(Traced.kt:46)
+ *   at arrow.core.continuations.Effect__TracingKt$traced$2.invoke(Traced.kt:46)
+ *   at arrow.core.continuations.Effect__FoldKt.fold(Fold.kt:92)
+ *   at arrow.core.continuations.Effect.fold(Unknown Source)
+ *   at MainKtKt.main(MainKt.kt:8)
+ *   at MainKtKt.main(MainKt.kt)
+ * ```
+ *
+ * NOTE:
+ * This implies a performance penalty of creating a stacktrace when calling [Raise.raise],
+ * but **this only occurs** when composing `traced`.
+ * The stacktrace creation is disabled if no `traced` calls are made within the function composition.
+ */
+@ExperimentalTraceApi
+public inline fun <R, A> Raise<R>.traced(
+  @BuilderInference program: Raise<R>.() -> A,
+  trace: (traced: Trace, R) -> Unit
+): A {
+  val itOuterTraced = this is DefaultRaise && isTraced
+  val nested = if (this is DefaultRaise && isTraced) this else DefaultRaise(true)
+  return try {
+    program.invoke(nested)
+  } catch (e: RaiseCancellationException) {
+    val r: R = e.raisedOrRethrow(nested)
+    trace(Trace(e), r)
+    if (itOuterTraced) throw e else raise(r)
+  }
+}
+
 /** Returns the raised value, rethrows the CancellationException if not our scope */
 @PublishedApi
 @Suppress("UNCHECKED_CAST")
 internal fun <R> CancellationException.raisedOrRethrow(raise: DefaultRaise): R =
-  if (this is RaiseCancellationException && this.raise === raise) raised as R
-  else throw this
+  when {
+    this is RaiseCancellationExceptionNoTrace && this.raise === raise -> raised as R
+    this is RaiseCancellationException && this.raise === raise -> raised as R
```

Comment:
```suggestion
    this.raise === raise && (this is RaiseCancellationExceptionNoTrace || this is RaiseCancellationException) -> raised as R
```

Reply:
This change introduces a compilation error because `raise` and `raised` properties don't exist in `CancellationException`. Those properties belong to `RaiseCancellationExceptionNoTrace` and `RaiseCancellationException`, so we need the smart cast here.